### PR TITLE
Order -> Contract Module Log

### PR DESCRIPTION
**Pull request by riadmen**
https://github.com/metasfresh/metasfresh/issues/15907

### CODE REVIEW

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SalesOrderLineModularContractHandler.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+	{
+		final I_C_Order order = orderBL.getById(OrderId.ofRepoId(orderLine.getC_Order_ID()));
+		if (SOTrx.ofBoolean(order.isSOTrx()).isPurchase())
+		{
+			return false;
+		}
+
+		final BPartnerLocationAndCaptureId orderBillPartnerId = orderBL.getBillToLocationId(order);
+
+		final boolean isMatchingPartner = Optional.ofNullable(WarehouseId.ofRepoIdOrNull(order.getM_Warehouse_ID()))
```
Comment (metas-rc) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SalesOrderLineModularContractHandler.java:115]: I don't think we need this validation. We can sell to any customer as long as the M_Warehouse.C_BPartner_ID is the one with the contract

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SalesOrderLineModularContractHandler.java [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+		}
+
+		final YearId harvestingYearId = YearId.ofRepoIdOrNull(order.getHarvesting_Year_ID());
+		if (harvestingYearId == null)
+		{
+			return false;
+		}
+
+		final ModularFlatrateTermRequest request = ModularFlatrateTermRequest.builder()
+				.bPartnerId(orderBillPartnerId.getBpartnerId())
```
Comment (metas-rc) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SalesOrderLineModularContractHandler.java:138]: warehousePartnerId

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SalesOrderLineModularContractHandler.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+				.map(FlatrateTermId::ofRepoId);
+	}
+
+	@Override
+	public void validateDocAction(final @NonNull I_C_OrderLine model, final ModularContractService.@NonNull ModelAction action)
+	{
+		switch (action)
+		{
+			case COMPLETED, VOIDED, REVERSED, REACTIVATED -> {}
+			case DELETED -> validateCanBeDeleted(model);
```
Comment (metas-rc) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SalesOrderLineModularContractHandler.java:237]: IMO DELETE is not a docAction.
I would move all the validation for the deleted line in a separate model interceptor, with 	``@ModelChange(timings = ModelValidator.TYPE_BEFORE_DELETE)``

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/interceptor/C_OrderLine.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+{
+	private final ModularContractService contractService;
+
+	public C_OrderLine(@NonNull final ModularContractService contractService)
+	{
+		this.contractService = contractService;
+	}
+
+	@ModelChange(timings = ModelValidator.TYPE_BEFORE_DELETE)
+	public void afterComplete(@NonNull final I_C_OrderLine orderLineRecord)
```
Comment (metas-rc) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/interceptor/C_OrderLine.java:45]: please, see comment above. Also, rename this method

**Review by metas-rc — CHANGES_REQUESTED** [review]
Please, check my comments

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SalesOrderLineModularContractHandler.java (hunk 1 above)
Comment (fp-ps) [reply backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SalesOrderLineModularContractHandler.java:115]: not needed actually. will fix it

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SalesOrderLineModularContractHandler.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+	public @NonNull Stream<FlatrateTermId> streamContractIds(@NonNull final I_C_OrderLine orderLine)
+	{
+		final I_C_Order order = orderBL.getById(OrderId.ofRepoId(orderLine.getC_Order_ID()));
+		final BPartnerLocationAndCaptureId orderBillPartnerId = orderBL.getBillToLocationId(order);
+		final YearId harvestingYearId = YearId.ofRepoIdOrNull(order.getHarvesting_Year_ID());
+
+		Check.assume(harvestingYearId != null, "Harvesting year ID should not be null at this stage!");
+
+		final ModularFlatrateTermRequest request = ModularFlatrateTermRequest.builder()
+				.bPartnerId(orderBillPartnerId.getBpartnerId())
```
Comment (metas-rc) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SalesOrderLineModularContractHandler.java:210]: warehouse partner ?

**Review by metas-rc — CHANGES_REQUESTED** [review]
Please, check small comment

**Review by metas-rc — APPROVED** [review]

**Pull request merged**